pkg/maps/tunnel/tunnel.go

```diff
@@ -17,6 +19,9 @@ const (
 
 	// MaxEntries is the maximum entries in the tunnel endpoint map
 	MaxEntries = 65536
+
+	TunnelEndpointIPv4 uint8 = 1
+	TunnelEndpointIPv6 uint8 = 2
```

Comment:
Is there an expectation that this values should match: https://github.com/cilium/cilium/blob/18513dbc1379a2d439163876e50dd68b009169fd/pkg/bpf/endpoint.go#L14-L18

Reply:
the key in endpoint_key originally was used for IPSEC encryption

https://github.com/cilium/cilium/blob/master/bpf/lib/encap.h#L189
https://github.com/cilium/cilium/blob/master/bpf/lib/encap.h#L215

```
                                                                                 
        tunnel = map_lookup_elem(&TUNNEL_MAP, key);                              
        if (!tunnel)                                                             
                return DROP_NO_TUNNEL_ENDPOINT;                                  
                                                                                 
#ifdef ENABLE_IPSEC                                                              
        if (tunnel->key) {                                                       
                __u8 min_encrypt_key = get_min_encrypt_key(tunnel->key);         
                                                                                 
                return encap_and_redirect_ipsec(ctx, tunnel->ip4,                
                                                min_encrypt_key,                 
                                                seclabel);                       
        }                                                                        
#endif                                                                     

```
since now we split the tunnel map key and value structure, we can remove the key from endpoint_key structure and keep the key in value tunnel_endpoint_info structure, what do you think?

Yes, I can remove TunnelEndpointIPv4/TunnelEndpointIPv6 and just use EndpointKeyIPv4/EndpointKeyIPv6, they are same.

Reply:
ah, I guess your original suggestion is to remove key from endpoint_key since it is not used anymore due to this PR, yes i will remove it.

Reply:
> ah, I guess your original suggestion is to remove key from endpoint_key since it is not used anymore due to this PR, yes i will remove it.

I was just curious. Not sure if we should remove it, but I think we should at least try to understand the code because now it is not clear to me what the purpose of `->key` is.

Reply:
> > ah, I guess your original suggestion is to remove key from endpoint_key since it is not used anymore due to this PR, yes i will remove it.
> 
> I was just curious. Not sure if we should remove it, but I think we should at least try to understand the code because now it is not clear to me what the purpose of `->key` is.

oh, I see. `->key` is used for IPSEC encryption, see 
https://github.com/cilium/cilium/blob/master/bpf/lib/encap.h#L189-L223
we can remove the `->key` from endpoint_key and keep it in tunnel_endpoint_info since `->key` is required for IPSEC, agree?

Reply:
@joestringer @kkourt actually the `-key` is used in endpoint_key `String()`
https://github.com/cilium/cilium/blob/master/pkg/bpf/endpoint.go#L68-L73
and endpoint_key `String()` is used in `DumpToMap`
https://github.com/cilium/cilium/blob/master/pkg/maps/lxcmap/lxcmap.go#L237-L245
`DumpToMap` is used in 
https://github.com/cilium/cilium/blob/master/daemon/cmd/datapath.go#L224-L267
so without fully understanding `DumpToMap`, to be on the safe side, I prefer not to remove `->key` from endpoint_key for this PR, what do you think?

Reply:
I think we just need to find all of the locations where this structure is used as a lookup key into the map, and those locations should tell us which fields are being populated in order to look up the values into the map; anything that is not being set there can be removed from the key version of the structure. Everything else can remain in the value.

Reply:
> I think we just need to find all of the locations where this structure is used as a lookup key into the map, and those locations should tell us which fields are being populated in order to look up the values into the map; anything that is not being set there can be removed from the key version of the structure. Everything else can remain in the value.

endpoint_key  `->key` is only used in endpoint_key String(), nowhere else used as part of map lookup 

```
func (k EndpointKey) String() string {
	if ip := k.ToIP(); ip != nil {
		return net.JoinHostPort(ip.String(), fmt.Sprintf("%d", k.Key))
	}
	return "nil"
}
```

I noticed above `k.Key` usage because I had compiling error about it after removing the `->key` from endpoint_key, I would have compiling errors for other places if the `->key` is used.  and above `return net.JoinHostPort(ip.String(), fmt.Sprintf("%d", k.Key))` is kind of strange to me, `net.JoinHostPort()` is suppose to join `host` and `port`, use `k.Key` as `port` does not make much sense to me, but I don't know the history of this code and why it is done that way. I actually tried to change `k.Key` to `key.Family` after removing `->key`  and tested in lab, seems no issue, so I have two options

1, remove the `->key` and replace above `k.Key` with `k.Family`
2, keep `->key` in endpoint_key

I prefer 2) unless someone know the history of above code and why use `k.Key` as `port` as part of the string generation, maybe `k.Key` has no any meaning here.

Reply:
I found following to commits:

```
commit 070214f089e78bf047ce671bd15d762174654ab6
Author: Tobias Klauser <tklauser@distanz.ch>
Date:   Wed Aug 26 17:25:02 2020 +0200

    Use net.JoinHostPort to construct network address strings
```
and 
```
commit b6989723a7cc0ebc07345d78d28b63bf3825f7ba
Author: John Fastabend <john.fastabend@gmail.com>
Date:   Tue Mar 12 10:18:40 2019 -0700

    cilium: ipsec, support rolling updates
    
```
@jrfastab @tobiaskohlbau 
is it safe to replace  `k.Key` with `k.Family` here,  we are considering removing the `->key` from endpoint_key structure since we are trying to separate tunnel map with key and value structure.

Reply:
Most likely this was a false ping and you would like to ping @tklauser.

Reply:
@jrfastab I went through the code change in `b6989723a7`, I believe it is safe to remove the `Key` from endpoint_key struct after this tunnel map key and value structure separation, @tklauser I also believe it is safe to replace `k.Key` with `key.Family` after reading `070214f0`, please let me know otherwise. @joestringer @kkourt , I will send a diff with `Key` removed from endpoint_key

Reply:
@joestringer @kkourt  I updated the PR with your comment addressed, Please let me know if I am miss anything, if it looks ok, Joe, could you trigger the CI again to test?